src/optim/FishLeg/fishleg.py

```diff
@@ -226,74 +393,113 @@ def update_aux(self) -> None:
         where :math:`\\theta` is the parameters of model, :math:`\lambda` is the
         auxliarary parameters.
         """
-        data_x, _ = next(iter(self.aux_dataloader))
-        data_x.to(self.device)
+        self.store_g = False
+        data = next(iter(self.aux_dataloader))
+        data = self._prepare_input(data)
 
         self.aux_opt.zero_grad()
         with torch.no_grad():
-            pred = self.draw(self.model, data_x)
+            samples = self.draw(self.model, data)
+
+        if True:
             g2 = 0.0
             for group in self.param_groups:
                 for p in group["params"]:
-                    g2 += p.grad.data.norm(p=2) ** 2
-            g2 = torch.sqrt(g2)
+                    grad = p.grad.data
+                    g2 = g2 + torch.sum(grad * grad)
+            g_norm = torch.sqrt(g2)
 
-        aux_loss = 0.0
-        for group in self.param_groups:
-            name = group["name"]
+        self.zero_grad()
+        self.nll(self.model, samples).backward()
 
-            grad_norm = [p.grad.data / g2 for p in group["params"]]
-            qg = group["Qv"](grad_norm)
+        reg_term = 0.0
+        quad_term = 0.0
+        linear_term = 0.0
 
-            for p, g, d_p, para_name in zip(
-                group["params"], grad_norm, qg, group["order"]
-            ):
+        for group in self.param_groups:
+            qg = group["Qv"]() if self.batch_speedup else group["Qv"](group["grad"])
 
-                self.plus_model._modules[name]._parameters[para_name] = (
-                    p.data + d_p * self.eps
-                )
-                self.minus_model._modules[name]._parameters[para_name] = (
-                    p.data - d_p * self.eps
-                )
+            for p, g, d_p in zip(group["params"], group["grad"], qg):
 
-                aux_loss -= 2 * torch.sum(g * d_p)
-                aux_loss += self.damping * d_p.norm(p=2) ** 2
+                grad = p.grad.data
+                quad_term = quad_term + torch.sum(grad * d_p)
+                linear_term = linear_term + torch.sum(g * d_p)
+                reg_term = reg_term + self.damping * torch.sum(d_p * d_p)
 
-        h_plus = self.nll(self.plus_model, data_x, pred)
-        h_minus = self.nll(self.minus_model, data_x, pred)
+        quad_term = quad_term**2
 
-        aux_loss += (h_plus + h_minus) / (self.eps**2)
-        aux_loss.backward()
+        aux_loss = 0.5 * (reg_term + quad_term) - linear_term
```

Comment:
When I run this in a training run, I see a negative loss every now and then. Is this what you also see?

Reply:
In my experiment of fine-tuning yes, if the model is pre-trained to a local minimum, loss is negative.

Reply:
OK, so what is the process for training from scratch? In this case I am still following the .step() update and the aux loss seems to be negative a lot during a training run.

Reply:
(the average over `g` is `-0.5 Trace(F^{-1} cov(g))`)

Reply:
Will damping be a big effect here? as `Qg` is actually `(F + \gamma I)^{-1}g`